### ISSUE TITLE
Multiple Devices Used

### ISSUE THREAD

**Issue by kjanoudi** [timelyart/Kairos#11]
**Describe the bug**

**To Reproduce**
1. I set 

```
user_data_directory = /tmp/kairos/user_data
share_user_data = true
```

2. I launch 5 instances of Kairos, all show the following:
```
2020-07-20 13:12:09.835 INFO tv - run: Running on a darwin operating system
2020-07-20 13:12:09.928 INFO tv - assign_user_data_directory: kairos_0 assigned
2020-07-20 13:12:12.089 INFO tv - check_driver: browser version: 84.0.4147.89
2020-07-20 13:12:12.090 INFO tv - check_driver: driver version: 84.0.4147.30
```

3. I cannot log into a 6th device, on the plan which allows 5 max devices. It disconnects me, and shows the screenshot below. Am i doing something wrong?

**Screenshots**
![image](https://user-images.githubusercontent.com/4011484/87966398-382d9e00-ca8b-11ea-9e50-28f034f80f63.png)

**Desktop (please complete the following information):**
 - OS: OSX
 - Browser: Chrome (latest)

**Additional context**

**Comment by timelyart** [comment]
1. Settings look good
2. & 3. What should happen is that you get:
`assign_user_data_directory: kairos_0 assigned`
`assign_user_data_directory: kairos_1 assigned`
`assign_user_data_directory: kairos_2 assigned`
`assign_user_data_directory: kairos_n assigned`

Are you running the commands with `sudo`? 
Furthermore, how much time is there between starting each instance? As part of the process Kairos needs to make a copy of the user data directory. 

In general, I don't own an OSX device so I cannot test and thus guarantee that this works on OSX. That being said, it should be largely the same as Linux. In fact, I know of several people who run it on OSX but they probably don't use that option.

**Comment by kjanoudi** [comment]
[[[Resolved]]]  - See comments below
 
It does not seem to matter how much time is spent between running the two, nor if I use sudo or not. Another issue I have seen, which may be related, is that once I launch an instance using an existing user data directory, 9/10 if i try to run the same command again, I see failures at the login screen, with a stacktrace from chromedriver. If i clear the directory using `rm -rf`, I am able to proceed just fine. Using the `-cls` flag does not impact this behavior in either way.

It always tries to use kairos_0. I am running them in different terminal tabs, using a single `./main.py` command in each one. Is that correct? I also noticed that it doesn't seem to actually look at the user data folder per say, it always goes a level up from there due to this:

```
    user_data_base_dir, tail = os.path.split(user_data_directory)
    kairos_data_directory = os.path.join(user_data_base_dir, 'kairos')
```

```
user_data_directory = /tmp/kairos/user_data
share_user_data = true

ls -al /tmp/kairos                                                                                                                                                        (07-20 17:04)
total 0
drwxr-xr-x   5 kjanoudi  wheel  160 Jul 20 17:03 .
drwxrwxrwt  13 root      wheel  416 Jul 20 13:04 ..
drwxr-xr-x   6 kjanoudi  wheel  192 Jul 20 17:01 kairos
drwxr-xr-x   6 kjanoudi  wheel  192 Jul 20 17:03 kairos_0
drwxr-xr-x   6 kjanoudi  wheel  192 Jul 20 17:01 user_data

ls -al /tmp/kairos/user_data                                                                                                                                   (07-20 17:09)
total 0
drwxr-xr-x  2 kjanoudi  wheel   64 Jul 20 17:07 .
drwxr-xr-x  5 kjanoudi  wheel  160 Jul 20 17:07 ..
```

> `python3 ./main.py -cls strategy.yaml`
```
2020-07-20 17:01:48.893 INFO main - clean_browser_data: kairos_0 removed
2020-07-20 17:01:48.893 INFO main - clean_browser_data: cleaning complete
2020-07-20 17:01:49.124 INFO tv - run: Running on a darwin operating system
2020-07-20 17:01:49.142 INFO tv - assign_user_data_directory: creating base user data directory 'kairos'. Please be patient while data is being copied ...
2020-07-20 17:01:51.069 INFO tv - check_driver: browser version: 84.0.4147.89
2020-07-20 17:01:51.069 INFO tv - check_driver: driver version: 84.0.4147.30
2020-07-20 17:01:51.072 INFO tv - create_browser: creating shared session for kairos user data directory
2020-07-20 17:01:59.606 INFO tv - login: Timeout exception
2020-07-20 17:01:59.607 INFO tv - login: About to click sign in
2020-07-20 17:01:59.694 INFO tv - login: About to click show_email_or_username
2020-07-20 17:02:10.087 ERROR tv - login: Message: 
Stacktrace:
0   chromedriver                        0x0000000105d48c49 chromedriver + 4893769
1   chromedriver                        0x0000000105ce20e3 chromedriver + 4473059
2   chromedriver                        0x00000001059558fd chromedriver + 751869
3   chromedriver                        0x00000001058c3c71 chromedriver + 154737
4   chromedriver                        0x00000001058eeb47 chromedriver + 330567
5   chromedriver                        0x00000001058e007d chromedriver + 270461
6   chromedriver                        0x00000001058ecbe8 chromedriver + 322536
7   chromedriver                        0x00000001058e0343 chromedriver + 271171
8   chromedriver                        0x00000001058b962a chromedriver + 112170
9   chromedriver                        0x00000001058ba635 chromedriver + 116277
10  chromedriver                        0x0000000105d0a5af chromedriver + 4638127
11  chromedriver                        0x0000000105d1791b chromedriver + 4692251
12  chromedriver                        0x0000000105d176bb chromedriver + 4691643
13  chromedriver                        0x0000000105cee109 chromedriver + 4522249
14  chromedriver                        0x0000000105d17ea3 chromedriver + 4693667
15  chromedriver                        0x0000000105d00073 chromedriver + 4595827
16  chromedriver                        0x0000000105d2f094 chromedriver + 4788372
17  chromedriver                        0x0000000105d4edb7 chromedriver + 4918711
18  libsystem_pthread.dylib             0x00007fff6ea7c109 _pthread_start + 148
19  libsystem_pthread.dylib             0x00007fff6ea77b8b thread_start + 15

2020-07-20 17:02:10.750 ERROR tv - snapshot: log/20200720_170210.png
```
> `python3 ./main.py strategy.yaml`

```

2020-07-20 17:03:54.531 INFO tv - run: Running on a darwin operating system
2020-07-20 17:03:54.532 INFO tv - create_database: Connection established to log/example.db
2020-07-20 17:03:54.549 INFO tv - assign_user_data_directory: creating user data directory 'kairos_0'. Please be patient while data is being copied ...
2020-07-20 17:03:54.849 INFO tv - assign_user_data_directory: kairos_0 assigned
2020-07-20 17:03:56.688 INFO tv - check_driver: browser version: 84.0.4147.89
2020-07-20 17:03:56.688 INFO tv - check_driver: driver version: 84.0.4147.30
2020-07-20 17:04:06.032 INFO tv - login: Timeout exception
2020-07-20 17:04:06.033 INFO tv - login: About to click sign in
2020-07-20 17:04:06.134 INFO tv - login: About to click show_email_or_username
2020-07-20 17:04:16.562 ERROR tv - login: Message: 
Stacktrace:
0   chromedriver                        0x000000010d2c5c49 chromedriver + 4893769
1   chromedriver                        0x000000010d25f0e3 chromedriver + 4473059
2   chromedriver                        0x000000010ced28fd chromedriver + 751869
3   chromedriver                        0x000000010ce40c71 chromedriver + 154737
4   chromedriver                        0x000000010ce6bb47 chromedriver + 330567
5   chromedriver                        0x000000010ce5d07d chromedriver + 270461
6   chromedriver                        0x000000010ce69be8 chromedriver + 322536
7   chromedriver                        0x000000010ce5d343 chromedriver + 271171
8   chromedriver                        0x000000010ce3662a chromedriver + 112170
9   chromedriver                        0x000000010ce37635 chromedriver + 116277
10  chromedriver                        0x000000010d2875af chromedriver + 4638127
11  chromedriver                        0x000000010d29491b chromedriver + 4692251
12  chromedriver                        0x000000010d2946bb chromedriver + 4691643
13  chromedriver                        0x000000010d26b109 chromedriver + 4522249
14  chromedriver                        0x000000010d294ea3 chromedriver + 4693667
15  chromedriver                        0x000000010d27d073 chromedriver + 4595827
16  chromedriver                        0x000000010d2ac094 chromedriver + 4788372
17  chromedriver                        0x000000010d2cbdb7 chromedriver + 4918711
18  libsystem_pthread.dylib             0x00007fff6ea7c109 _pthread_start + 148
19  libsystem_pthread.dylib             0x00007fff6ea77b8b thread_start + 15

2020-07-20 17:04:17.262 ERROR tv - snapshot: log/20200720_170416.png

========================================
2020-07-20 17:04:17 - End Program
Elapsed time: 23 seconds
========================================
```

> `rm -rf /tmp/kairos/kairos`
> `rm -rf /tmp/kairos/kairos_0`
> `python3 ./main.py strategy.yaml`

```
2020-07-20 17:07:40.587 INFO tv - run: Running on a darwin operating system
2020-07-20 17:07:40.604 INFO tv - assign_user_data_directory: creating base user data directory 'kairos'. Please be patient while data is being copied ...
2020-07-20 17:07:42.527 INFO tv - check_driver: browser version: 84.0.4147.89
2020-07-20 17:07:42.528 INFO tv - check_driver: driver version: 84.0.4147.30
2020-07-20 17:07:42.531 INFO tv - create_browser: creating shared session for kairos user data directory
2020-07-20 17:07:54.389 INFO tv - login: Timeout exception
2020-07-20 17:07:54.390 INFO tv - login: About to click sign in
2020-07-20 17:07:54.479 INFO tv - login: About to click show_email_or_username
2020-07-20 17:07:55.343 INFO tv - login: filling username
2020-07-20 17:07:55.462 INFO tv - login: filling password
2020-07-20 17:07:55.462 INFO tv - login: typing them in
2020-07-20 17:07:55.852 INFO tv - login: clicking log in
2020-07-20 17:07:56.986 INFO tv - login: logged in successfully at tradingview.com
2020-07-20 17:07:59.483 INFO tv - close_cookies_message: Cookie banner not found
2020-07-20 17:07:59.483 INFO tv - destroy_browser: Closing connection to sql
2020-07-20 17:07:59.641 INFO tv - create_browser: restarting kairos ... 
2020-07-20 17:07:59.642 INFO tv - assign_user_data_directory: creating user data directory 'kairos_0'. Please be patient while data is being copied ...
2020-07-20 17:07:59.719 INFO tv - assign_user_data_directory: kairos_0 assigned
2020-07-20 17:08:01.505 INFO tv - check_driver: browser version: 84.0.4147.89
2020-07-20 17:08:01.506 INFO tv - check_driver: driver version: 84.0.4147.30
2020-07-20 17:08:10.162 INFO tv - login: Timeout exception
2020-07-20 17:08:10.162 INFO tv - login: About to click sign in
2020-07-20 17:08:10.242 INFO tv - login: About to click show_email_or_username
2020-07-20 17:08:11.588 INFO tv - login: filling username
2020-07-20 17:08:11.652 INFO tv - login: filling password
2020-07-20 17:08:11.653 INFO tv - login: typing them in
2020-07-20 17:08:11.874 INFO tv - login: clicking log in
2020-07-20 17:08:12.822 INFO tv - login: logged in successfully at tradingview.com
```

Screenshot created by the error handler:
![image](https://user-images.githubusercontent.com/4011484/87989614-59ed4c00-cab0-11ea-914e-12a69f827bf3.png)

**Comment by kjanoudi** [comment]
I believe I found the issue:

https://github.com/timelyart/Kairos/blob/master/kairos/tools.py#L310

Changing to 
`if process.name().lower().find('chrome') >= 0:`

seems to set the directories fine.

Login bug is due to a change that sometimes shows the username/password fields by default, causing this selector to hang:

https://github.com/timelyart/Kairos/blob/master/tv/tv.py#L1801

```
if not show_email_or_username.get_attribute('class').index('i-hidden'):
            wait_and_click(browser, css_selectors['show_email_or_username'])
```

Resolved my issue with the logins above

I may submit PR later on, but in case i don't, fixes are above

**Comment by timelyart** [comment]
Thank you for debugging. Can you confirm that parallel runs are working as intended now?

> Login bug is due to a change that sometimes shows the username/password fields by default, causing this selector to hang:

This is strange. I haven't seen this issue before. Are you sure that the Cookie banner wasn't in the way? TV changed the DOM so  automatically accepting the cookies isn't working right now. Just to make sure, when the login failed, the cookie banner was outside the area of the username/password fields, i.e. the Cookie banner wasn't hiding those fields by laying on top of it?

> if not show_email_or_username.get_attribute('class').index('i-hidden'):
            wait_and_click(browser, css_selectors['show_email_or_username'])

I don't see how this can work, since **show_email_or_username** isn't an object yet. I suspect, you forgot to post the initialization of the element.

**Comment by kjanoudi** [comment]
Strangely, I never get the cookie banner. I believe that may be because they only present it to Europe, and I'm in USA. 

`2020-07-21 11:44:41.688 INFO tv - close_cookies_message: Cookie banner not found` - every time

But I did confirm several times by running in foreground that there are times where the `show_email_or_username` is hidden, via the `i-hidden` attribute. You're correct that the code I posted was not completely working - I'm rather new at python. This is what ended up working in both cases for me.

```
show_email_or_username = find_element(browser, css_selectors['show_email_or_username'])
        if show_email_or_username.get_attribute('class').find('i-hidden') < 0:
            wait_and_click(browser, css_selectors['show_email_or_username'])
```
